miner/worker.go

```diff
@@ -96,677 +58,181 @@ type environment struct {
-				continue
-			}
-			// Different block could share same sealhash, deep copy here to prevent write-write conflict.
-			var (
-				receipts = make([]*types.Receipt, len(task.receipts))
-				logs     []*types.Log
-			)
-			for i, taskReceipt := range task.receipts {
-				receipt := new(types.Receipt)
-				receipts[i] = receipt
-				*receipt = *taskReceipt
-
-				// add block location fields
-				receipt.BlockHash = hash
-				receipt.BlockNumber = block.Number()
-				receipt.TransactionIndex = uint(i)
-
-				// Update the block hash in all logs since it is now available and not when the
-				// receipt/log of individual transactions were created.
-				receipt.Logs = make([]*types.Log, len(taskReceipt.Logs))
-				for i, taskLog := range taskReceipt.Logs {
-					log := new(types.Log)
-					receipt.Logs[i] = log
-					*log = *taskLog
-					log.BlockHash = hash
-				}
-				logs = append(logs, receipt.Logs...)
-			}
-			// Commit block and state to database.
-			_, err := w.chain.WriteBlockAndSetHead(block, receipts, logs, task.state, true)
-			if err != nil {
-				log.Error("Failed writing block to chain", "err", err)
-				continue
-			}
-			log.Info("Successfully sealed new block", "number", block.Number(), "sealhash", sealhash, "hash", hash,
-				"elapsed", common.PrettyDuration(time.Since(task.createdAt)))
-
-			// Broadcast the block and announce chain insertion event
-			w.mux.Post(core.NewMinedBlockEvent{Block: block})
-
-		case <-w.exitCh:
-			return
-		}
+	// Run the consensus preparation with the default or customized consensus engine.
+	if err := miner.engine.Prepare(miner.chain, header); err != nil {
+		log.Error("Failed to prepare header for sealing", "err", err)
+		return nil, err
+	}
+	// Could potentially happen if starting to mine in an odd state.
+	// Note genParams.coinbase can be different with header.Coinbase
+	// since clique algorithm can modify the coinbase field in header.
+	env, err := miner.makeEnv(parent, header, genParams.coinbase)
+	if err != nil {
+		log.Error("Failed to create sealing context", "err", err)
+		return nil, err
+	}
+	if header.ParentBeaconRoot != nil {
+		context := core.NewEVMBlockContext(header, miner.chain, nil)
+		vmenv := vm.NewEVM(context, vm.TxContext{}, env.state, miner.chainConfig, vm.Config{})
+		core.ProcessBeaconBlockRoot(*header.ParentBeaconRoot, vmenv, env.state)
 	}
+	return env, nil
 }
 
 // makeEnv creates a new environment for the sealing block.
-func (w *worker) makeEnv(parent *types.Header, header *types.Header, coinbase common.Address) (*environment, error) {
+func (miner *Miner) makeEnv(parent *types.Header, header *types.Header, coinbase common.Address) (*environment, error) {
 	// Retrieve the parent state to execute on top and start a prefetcher for
 	// the miner to speed block sealing up a bit.
-	state, err := w.chain.StateAt(parent.Root)
+	state, err := miner.chain.StateAt(parent.Root)
```

Comment:
Originally we spin up the prefetcher but now it's dropped, any particular reason?

It's fine if there are some reasons behind, as we might just a bit less performant without prefetcher.

```go
diff --git a/miner/worker.go b/miner/worker.go
index 7e038b0f30..675c6652b4 100644
--- a/miner/worker.go
+++ b/miner/worker.go
@@ -93,6 +93,8 @@ func (miner *Miner) generateWork(params *generateParams) *newPayloadResult {
        if err != nil {
                return &newPayloadResult{err: err}
        }
+       defer work.state.StopPrefetcher()
+
        if !params.noTxs {
                interrupt := new(atomic.Int32)
                timer := time.AfterFunc(miner.config.Recommit, func() {
@@ -209,6 +211,8 @@ func (miner *Miner) makeEnv(parent *types.Header, header *types.Header, coinbase
        if err != nil {
                return nil, err
        }
+       state.StartPrefetcher("miner")
+
        // Note the passed coinbase may be different with header.Coinbase.
        return &environment{
                signer:   types.MakeSigner(miner.chainConfig, header.Number, header.Time),

```

Reply:
Yes, I recently did some benchmarks with the prefetcher enabled and disabled and it turns out the prefetcher in the miner did not influence block verification that much. Also it accounted for a lot of unnecessary disk i/o and cpu, see also: https://github.com/ethereum/go-ethereum/pull/28440

Cpu actually almost halfed in that experiment and block verification was faster, even though snapshot reads were slightly slower